File: examples/java/src/main/java/org/wickra/gym/examples/Rollout.java
// A runnable Java example: load the momentum_discrete spec and its candle
// dataset over the wickra-gym C ABI binding, then drive a fixed long policy and
// print each step. The rollout is byte-identical to the other language examples
// on the same seed.
//
//   cargo build --release -p wickra-gym-c
//   (cd bindings/java && mvn -q install -DskipTests)
//   mvn -q -f examples/java exec:exec
package org.wickra.gym.examples;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.wickra.gym.Env;

public final class Rollout {
    private static final Pattern REWARD = Pattern.compile("\"reward\":(-?[0-9.eE+-]+)");
    private static final Pattern TERMINATED = Pattern.compile("\"terminated\":(true|false)");
    private static final Pattern TRUNCATED = Pattern.compile("\"truncated\":(true|false)");

    private Rollout() {
    }

    private static Path dataDir() {
        Path dir = Path.of(System.getProperty("user.dir")).toAbsolutePath();
        for (int i = 0; i < 10 && dir != null; i++) {
            Path candidate = dir.resolve("examples").resolve("data");
            if (Files.isDirectory(candidate)) {
                return candidate;
            }
            dir = dir.getParent();
        }
        throw new IllegalStateException("examples/data not found");
    }

    private static double group1Double(Pattern p, String s) {
        Matcher m = p.matcher(s);
        if (!m.find()) {
            throw new IllegalStateException("missing field in: " + s);
        }
        return Double.parseDouble(m.group(1));
    }

    private static boolean group1Bool(Pattern p, String s) {
        Matcher m = p.matcher(s);
        if (!m.find()) {
            throw new IllegalStateException("missing field in: " + s);
        }
        return Boolean.parseBoolean(m.group(1));
    }

    public static void main(String[] args) throws IOException {
        Path data = dataDir();
        String spec = Files.readString(data.resolve("specs").resolve("momentum_discrete.json"));
        String candles = Files.readString(data.resolve("candles.json"));

        try (Env env = new Env(spec)) {
            env.command("{\"cmd\":\"load\",\"candles\":" + candles + "}");
            System.out.println("reset: " + env.command("{\"cmd\":\"reset\",\"seed\":42}"));

            double equity = 0.0;
            for (int step = 0; ; step++) {
                String result = env.command("{\"cmd\":\"step\",\"action\":2}");
                double reward = group1Double(REWARD, result);
                boolean terminated = group1Bool(TERMINATED, result);
                boolean truncated = group1Bool(TRUNCATED, result);
                equity += reward;
                System.out.printf(
                        "step %d: reward %+.6f  equity %+.6f  terminated=%b truncated=%b%n",
                        step, reward, equity, terminated, truncated);
                if (terminated || truncated) {
                    break;
                }
            }
        }
    }
}

File: examples/c/rollout.cpp
// A minimal C++ example: load a spec and its candle dataset, then drive a fixed
// long policy and print each step. DATA_DIR is injected by CMake.
//
// This goes through `wickra_gym.hpp`, the C++ hull shipped beside the C header,
// because that hull is what a C++ caller is meant to use: it owns and frees the
// handle, runs the two-call length protocol behind `wickra_gym_command` for you
// -- the core carries the produced-but-undelivered response between the two
// calls, so a mutating `step` runs once, not twice -- and turns a refusal into
// an exception rather than a negative integer that is easy to ignore. Calling
// the C functions directly from C++ works too, but then the hull would be
// shipped without anything building it, which is how it came to reference a
// type that does not exist.
#include <cstdio>
#include <fstream>
#include <sstream>
#include <string>

#include "wickra_gym.hpp"

static std::string slurp(const std::string &path) {
    std::ifstream f(path, std::ios::binary);
    std::ostringstream ss;
    ss << f.rdbuf();
    return ss.str();
}

int main() {
    const std::string dir = DATA_DIR;
    const std::string spec = slurp(dir + "/specs/momentum_discrete.json");
    const std::string candles = slurp(dir + "/candles.json");
    if (spec.empty() || candles.empty()) {
        std::fprintf(stderr, "cannot read example data\n");
        return 1;
    }

    int steps = 0;
    try {
        wickra::Env env(spec);
        env.command("{\"cmd\":\"load\",\"candles\":" + candles + "}");
        std::printf("wickra-gym %s\n", wickra::Env::version().c_str());
        std::printf("reset: %s\n", env.command("{\"cmd\":\"reset\",\"seed\":42}").c_str());

        for (;;) {
            const std::string step = env.command("{\"cmd\":\"step\",\"action\":2}");
            // An in-band refusal: the ABI answered, the core declined. That is a
            // response rather than an error, so the hull does not throw on it.
            if (step.find("\"ok\":false") != std::string::npos) {
                std::fprintf(stderr, "the environment refused a step: %s\n", step.c_str());
                return 1;
            }
            std::printf("step %d: %s\n", steps, step.c_str());
            const bool done = step.find("\"terminated\":true") != std::string::npos ||
                              step.find("\"truncated\":true") != std::string::npos;
            ++steps;
            if (done) {
                break;
            }
        }
    } catch (const wickra::GymError &err) {
        // Every failure arrives here: a spec the core rejects, a command it does
        // not know, a response that changed length between the two ABI calls.
        std::fprintf(stderr, "%s\n", err.what());
        return 1;
    }

    if (steps == 0) {
        std::fprintf(stderr, "the episode ended before its first step\n");
        return 1;
    }
    return 0;
}
